### PR TITLE
Reject invalid MPEG puts for certain lib versions

### LINKED ISSUE #3318: Stuck issue with Mega Man Maverick Hunter X
Reported form http://forums.ppsspp.org/showthread.php?tid=384&pid=41935#pid41935
when run this game it freezes at ACCESS page, press start button can continue it.

### LINKED ISSUE #6657: Go! Sudoku - random lag
Game drops down to a very low framerate during gameplay.
Happens sometimes
- when correctly filling up a box before anything else
- randomly

### PR DESCRIPTION
**Pull request by unknownbrackets**
The good news: this reportedly fixes #3318, and fixes #6657.

This change is probably a bit risky, but I've tested a number of games and haven't run into any video problems.  I still don't really know what "PMP videos" are and have not tested the specialized code dedicated to them.

Here are the major things this does:
1. Prevents searching for a startcode beyond available data.  This could've been causing corrupted audio frames in some cases.
2. More strictly and more immediately demuxes MPEG frames.  This may also help corrupted audio frames, but might also break somewhere it was accidentally working.
3. Exposes the MPEG write position to the game in an actual ring.  This is how it's really supposed to work anyway, and might make it easier to decode more directly from PSP RAM some day.
4. Reduces the required mpeg memory for older libs, per tests.  Might matter for some memory allocation.
5. Advances packetsAvail incrementally (required for #3318), rather than resetting it after decode.  I played this one safe, and only did it for the necessary lib versions.

However, one note: I have found two 1.03 mpeg.prx's that have different behavior.  Technically, some of this behavior is correct for 1.03a but not for 1.03b.  Hoping it doesn't matter, I don't want to go down the route of crcing them for version numbers...

-[Unknown]

### CODE REVIEW

**Comment by zminhquanz** [comment]
Are you plan to write UMD video in the future ?

**Comment by unknownbrackets** [comment]
Well, it'd probably be more worthwhile to ask FFmpeg to detect and support atrac3plus (which it can already decode) in standard video files.  If they did that, most likely UMD videos would be playable in just any video player you wanted, as long as it used ffmpeg.

That sounds like a better outcome than PPSSPP having support for it, especially because PPSSPP is a pretty terrible video player.  It can't handle subtitles or fonts in the stream, doesn't have fancy scrubbing controls, doesn't support embedded album art, doesn't add latency to video when audio is lagging (which would be a terrible thing for an emulator but is a great thing for a video player), etc.

Basically, there are a bunch of really popular video players out there that people have devoted hundreds or even thousands of hours to making great.  And they are great.  I feel no motivation to throw hundreds of my own hours into making a not-as-great video player.

Isn't it better for a great video player to support another video format, rather than a great emulator to try to "me too" being a video player?

-[Unknown]

**Comment by daniel229** [comment]
Also fixes https://github.com/hrydgard/ppsspp/issues/6657

**Comment by unknownbrackets** [comment]
Oh, cool.  I guess that explains what was happening there.  Does it help #8526?

-[Unknown]

**Comment by daniel229** [comment]
No.

**Comment by unknownbrackets** [comment]
@zminhquanz btw, here's the ticket for that: https://trac.ffmpeg.org/ticket/3233

-[Unknown]

**Comment by zminhquanz** [comment]
oh , i'm very pleasure , good job

**Comment by hrydgard** [comment]
No, that's not related.
